File: problems/29/solution.go
func divide(dividend int, divisor int) int {
    neg, ans := 0, 0
    
    if dividend == -2147483648 && divisor == -1 {
        return abs(dividend) - 1
    }
    
    if (dividend >= 0 && divisor > 0) || (dividend <= 0 && divisor < 0) {
        neg = 1
    } else {
        neg = -1
    }
    
    dividend, divisor = abs(dividend), abs(divisor)
    
    if divisor == 1 {
        ans = dividend
    } else {
        for dividend >= divisor {
            dividend -= divisor
            ans++
        }
    }
    
    return ans * neg
}

func abs(a int) int {
    if a < 0 {
        return a * -1
    }
    
    return a
}

File: problems/29/solution.java
class Solution {
    public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }
        
        int neg = 2, ans = 0, HALF_INT_MIN = Integer.MIN_VALUE / 2;

        if (dividend > 0) {
            neg--;
            dividend = -dividend;
        }
        
        if (divisor > 0) {
            neg--;
            divisor = -divisor;
        }
        
        while (dividend <= divisor) {
            int powerOfTwo = -1, value = divisor;
            
            while (value >= HALF_INT_MIN && value + value >= dividend) {
                value += value;
                powerOfTwo += powerOfTwo;
            }
            
            ans += powerOfTwo;
            dividend -= value;
        }

        if (neg != 1) {
            return -ans;
        }
        
        return ans;
    }
}